kvclient: separate out proxy sending

andrewbaptist opened the pull request:
Previously in DistSender, a proxy request was handled using the same logic within sendPartialBatch and sendToReplicas. There were short circuits to handle the different cases of retries, but in cases with changing range boundaries, it could return the wrong error. This change intercepts proxy request at the start of DistSender.Send and sends to the transport bypassing the rest of the DistSender retry logic. This simplifies the code for proxy requests.


Epic: none
Fixes: #123965
Informs: #123146

Release note: None

Commit:
kvclient: separate out proxy sending

Previously in DistSender, a proxy request was handled using the same
logic within sendPartialBatch and sendToReplicas. There were short
circuits to handle the different cases of retries, but in cases with
changing range boundaries, it could return the wrong error. This change
intercepts proxy request at the start of DistSender.Send and sends to
the transport bypassing the rest of the DistSender retry logic. This
simplifies the code for proxy requests.

Epic: none
Fixes: #123965
Informs: #123146

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124066)
<!-- Reviewable:end -->

andrewbaptist commented:
Note the only test failure was on TestProtectedTimestampManagement which is unrelated and tracked as #124167

nvb reviewed (commented):
I like the simplicity of this new approach.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-:-O-3CCsB25QclrPwWONP:bdr6ap6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3AgBz9lSrjCDxb-YX:-O-3AgBz9lSrjCDxb-YY:b-js75kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1311)):*
> The proxy node inserts any range information

Can we update this comment to reflect the conditional nature of this insert? We won't replace newer information with stale information, will we?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3ArVZ-BgTzW8PUeZL:-O-3ArVZ-BgTzW8PUeZM:blacrt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1312)):*
> but doesn't use it for routing decisions

It would be useful to give a few examples of what this means in practice. For instance, if the proxy node knows that the range has been split, it will still send the request. Similarly, if it knows that the lease has been moved, it will still send to the proxy-requested leaseholder.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3C4Wx4_8QtYT7LeOJ:-O-3C4Wx4_8QtYT7LeOK:b-5nkro8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1312)):*
> ```Go
> // information, the client will update its cache and routing information and
> // retry locally. The proxy node inserts any range information but doesn't use
> // it for routing decisions.
> ```

Do we have any tests that demonstrate the change in behavior for proxying that this change is making? If not, should we?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3BkOI-pRHlk0By63H:-O-3BkOI-pRHlk0By63I:b64yua8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1335)):*
> ```Go
> 	requestToSend.ProxyRangeInfo = nil
> 	requestToSend.Replica = transport.NextReplica()
> 	br, err := transport.SendNext(ctx, requestToSend)
> ```

Should we be consulting dist sender circuit breakers here, like we do in `sendToReplicas`?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1340 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3BzL15YIH79J1JJfs:-O-3BzL2EV8Lyzegw-ZL:b-7n4o9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1340)):*
> ```Go
> 	if err != nil {
> 		// Convert the error from an error on the transport to a
> 		// ProxyFailedError This can then be sent over the wire and decoded by
> ```

"ProxyFailedError This"

missing punctuation.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2045 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3ARhmAvB5dyMQdiwh:-O-3ARhmAvB5dyMQdiwi:b-5wf4ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L2045)):*
> ```Go
> // a sendError. Node.maybeProxyRequest specifically excludes this error from
> // propagation over the wire and instead return the NLHE from local evaluation.
> var ProxyFailedWithSendError = kvpb.NewErrorf("proxy request failed with send error")
> ```

It looks like this is no longer used. Is that intentional? If so, can be remove it?


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
Thanks - It would have been easier if I had done this first, but I'm glad this worked out this way.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-:-O-Xr69e4iLiSUsX1JnK:byvv65w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3AgBz9lSrjCDxb-YX:-O-XmIpJ78anO_nRxbOA:bj6xa62) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1311)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> The proxy node inserts any range information

Can we update this comment to reflect the conditional nature of this insert? We won't replace newer information with stale information, will we?
</blockquote></details>

I added some clarification.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3ArVZ-BgTzW8PUeZL:-O-XnC8EBOFMmRWvymDV:b-p8epz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1312)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> but doesn't use it for routing decisions

It would be useful to give a few examples of what this means in practice. For instance, if the proxy node knows that the range has been split, it will still send the request. Similarly, if it knows that the lease has been moved, it will still send to the proxy-requested leaseholder.
</blockquote></details>

I added more context here. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3C4Wx4_8QtYT7LeOJ:-O-XttGt4IJvR4LUnEu2:b-6680xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1312)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we have any tests that demonstrate the change in behavior for proxying that this change is making? If not, should we?
</blockquote></details>

Yes we should - let me try and create a test for this. I am going to focus on just the case where the proxy node has newer information that includes a split and validate the request is not split.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3BkOI-pRHlk0By63H:-O-Xq1NX2ixJ21J15IX_:b1yivgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1335)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we be consulting dist sender circuit breakers here, like we do in `sendToReplicas`?
</blockquote></details>

Good catch. I added dist sender tracking. I didn't pull all of the logic, and I added a TODO to refactor this a little bit. There is a special case that the normal logic handles related to "local" requests since they don't cancel the same way. This doesn't apply to proxy requests.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1340 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3BzL15YIH79J1JJfs:-O-Xqyg_1-3r5OgHN8cm:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1340)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"ProxyFailedError This"

missing punctuation.
</blockquote></details>

Fixed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2045 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3ARhmAvB5dyMQdiwh:-O-Xr4rB8ZUx5M8SInI3:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L2045)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It looks like this is no longer used. Is that intentional? If so, can be remove it?
</blockquote></details>

Removed


<!-- Sent from Reviewable.io -->

Commit:
kvclient: verify a proxy request is not split

If a proxy request is against a range which is subsequently split, the
request should still follow what the original client requested rather
than the updated cache. This prevents complex errors from split requests
as they can't be processed correctly.

Informs: #123965

Release note: None

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124066)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3C4Wx4_8QtYT7LeOJ:-O-byoAD6JlUpZGr0Rr5:bhijt75) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1312)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Yes we should - let me try and create a test for this. I am going to focus on just the case where the proxy node has newer information that includes a split and validate the request is not split.
</blockquote></details>

I added a test for this case. I put it as a separate commit for now to validate that it indeed reproduces the issue (with an error because of different proxy results). I can squash the commits or leave them separate.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 2 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124066)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3AgBz9lSrjCDxb-YX:-O-fkcyr-1l9uWiCx4CC:b-d9lkwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1311)):*
> The proxy node inserts any newer range information but doesn't use it for routing decisions

The meaning of "it" is ambiguous here. Consider splitting into:
> The request carries range information, which may be newer, older, or the same as that in the proxy node's range cache. If the request carries newer information than what is known on the proxy node, the proxy node's cache is updated. However, even if the request carries old information, the proxy node's cache is never used to inform routing decisions.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3BkOI-pRHlk0By63H:-O-fm8t3E-ZCw4WrUrc6:b-u1ufm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1335)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Good catch. I added dist sender tracking. I didn't pull all of the logic, and I added a TODO to refactor this a little bit. There is a special case that the normal logic handles related to "local" requests since they don't cancel the same way. This doesn't apply to proxy requests.
</blockquote></details>

I'm confused how we're able to get away with not calling `replicaCircuitBreakerToken.Done`. Aren't we tracking state in the `ReplicaCircuitBreaker` that requires a Done call, like `inflightReqs`?

Also, which special case are you referring to, and why does it not apply to proxy requests?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1313 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-flZov1_EW8kTdRvxn:-O-flZov1_EW8kTdRvxo:bz2ik03) ([raw file](https://github.com/cockroachdb/cockroach/blob/38e04614f782b50f08d7cda6b4afac18d2e6d4e4/pkg/kv/kvclient/kvcoord/dist_sender.go#L1313)):*
> ```Go
> // retry locally. The proxy node inserts any newer range information but doesn't
> // use it for routing decisions, outated information is ignored. The transport
> // is built used based on the information in the request not the local cache. As
> ```

"build used based"
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 6307 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-fjhPEA9PcUBA4bvD-:-O-fjhPEA9PcUBA4bvD0:b-n9y941) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9bf3abaac3711d7137dde811856dfefdc55711f/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L6307)):*
> ```Go
> 
> // TestProxyRequestDoesNotSplit tests that a proxy request does not split even
> // if the proxy node would normally split the request . Spliting the request can
> ```

nit: remove space after request.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 6361 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-fjzOf4npPQR6vNZps:-O-fjzOf4npPQR6vNZpt:b-sidyi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9bf3abaac3711d7137dde811856dfefdc55711f/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L6361)):*
> ```Go
> 
> 	// Track how many times the transport is called.
> 	sendCount := atomic.Int64{}
> ```

Consider splitting `sendCount` into a count of non-proxied requests and proxied requests (`var sendCount, proxyCount atomic.Int32`). That way, you'll have `sendCount.Load() == 0` in the main part of this test, which would avoid confused by readers.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-:-O-keczE-8QGnIAo9dG_:b-9qdoz3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3AgBz9lSrjCDxb-YX:-O-kRQXQ2XsaJnU47jOP:b1zf2ks) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1311)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> The proxy node inserts any newer range information but doesn't use it for routing decisions

The meaning of "it" is ambiguous here. Consider splitting into:
> The request carries range information, which may be newer, older, or the same as that in the proxy node's range cache. If the request carries newer information than what is known on the proxy node, the proxy node's cache is updated. However, even if the request carries old information, the proxy node's cache is never used to inform routing decisions.
</blockquote></details>

Thanks, I rewrite this entire comment, so please take one more look through it. There is a lot of subtly here so I agree its important to make it clear.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-3BkOI-pRHlk0By63H:-O-kWREl6LmihVkHkaNL:bhr7sgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a97a68f0cfb47c3f5c962a4b4d0831ea7bf44151/pkg/kv/kvclient/kvcoord/dist_sender.go#L1335)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm confused how we're able to get away with not calling `replicaCircuitBreakerToken.Done`. Aren't we tracking state in the `ReplicaCircuitBreaker` that requires a Done call, like `inflightReqs`?

Also, which special case are you referring to, and why does it not apply to proxy requests?
</blockquote></details>

You're right! I do need to call Done here to prevent a leak. This code is somewhat confusing because of the way `transport.SendNext` handles cancellation. I originally thought that Done was only necessary to handle the difference between local and remote evaluation, but is also needed to prevent the request leak. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1313 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-flZov1_EW8kTdRvxn:-O-kLy-ZBRPbTVXsP2x7:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/38e04614f782b50f08d7cda6b4afac18d2e6d4e4/pkg/kv/kvclient/kvcoord/dist_sender.go#L1313)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"build used based"
</blockquote></details>

Fixed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 6307 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-fjhPEA9PcUBA4bvD-:-O-kb_kW7f4LOEaNTx4O:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9bf3abaac3711d7137dde811856dfefdc55711f/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L6307)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: remove space after request.
</blockquote></details>

Done
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 6361 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-O-fjzOf4npPQR6vNZps:-O-kcTLB-wjiwuOOEwwj:b-zfx2tu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9bf3abaac3711d7137dde811856dfefdc55711f/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L6361)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider splitting `sendCount` into a count of non-proxied requests and proxied requests (`var sendCount, proxyCount atomic.Int32`). That way, you'll have `sendCount.Load() == 0` in the main part of this test, which would avoid confused by readers.
</blockquote></details>

I made the reset explicit. It is hard to cleanly determine in the transport whether it is a proxy request or not and rather than trying to parse it out, it is simpler and more resilient to explicitly reset it. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124066#-:-O-rDlgu0fIKJ0EwAe1o:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->

andrewbaptist commented:
bors r=nvanbenschoten

andrewbaptist commented:
bors+

andrewbaptist commented:
bors retry
